src/ILLink.Tasks/LinkTask.cs

```diff
@@ -49,17 +52,20 @@ public class ILLink : ToolTask
 		///   A list of XML root descriptor files specifying linker
 		///   roots at a granular level. See the mono/linker
 		///   documentation for details about the format.
+		///   Maps to '-x'.
 		/// </summary>
 		public ITaskItem [] RootDescriptorFiles { get; set; }
 
 		/// <summary>
 		///   Boolean specifying whether to clear initlocals flag on methods.
+		///   Maps to '--enable-opt clearinitlocals'.
 		/// </summary>
 		public bool ClearInitLocals { get; set; }
```

Comment:
Perhaps we should re-think about this and do all optimizations consistently not just this one

Reply:
Two options come to mind:
We could expose a bool for each optimization, with a default value of `true` for the default optimizations (I would add a test to check that these stay in sync). It would be used like this:
```xml
<ILLink ClearInitLocals="true" BeforeFieldInit="false" ... />
```
This would be my recommendation - to me it feels more like the "msbuild" way to pass parameters to tasks.

Another option would be to specify enabled/disabled optimizations as a semicolon-separated string (or maybe as separate strings in an ItemGroup), used like this:
```xml
<ILLink EnableOptimizations="clearinitlocals;sealer" DisableOptimizations="beforefieldinit;ipconstprop" ... />
```
which has the advantage that we don't need to add more task parameters for new optimizations, but doesn't restrict the strings that can be passed in to the task, which IMO is an unusual way to expose feature-level functionality. This kind of pattern is more often used for things like `DefineConstants` or inputs that take lists of arbitrary file paths.

Reply:
```
<ILLink ClearInitLocals="true" BeforeFieldInit="false" ... />
```

sounds good to me but where do you specify the assemblies the optimizations will apply to? Allowing SDKs to enable e.g. sealer for their own assembly only is something we need to support

Reply:
For per-assembly settings, we can choose:
1. A per-optimization input (ItemGroup of assembly names)
```xml
<ItemGroup>
   <_ClearInitLocalsAssemblies Include="Asm1;Asm2" />
</ItemGroup>
<ILLink ClearInitLocalsAssemblies="@(_ClearInitLocalsAssemblies)" />
```
2. A single list of assemblies with metadata in the form of strings
(here I'm combining it with the `AssemblyPaths` input, and the task would just take out the filename without extension to use it as the assembly name - or we could use a separate ItemGroup input):
```xml
<ItemGroup>
  <_AssemblyPaths Include="path/to/Asm1.dll">
    <enabledoptimizations>clearinitlocals;...</enabledoptimizations>
    <disabledoptimizations>beforefieldinit;...</disabledoptimizations>
  </_AssemblyPaths>
</ItemGroup>
<ILLink AssemblyPaths="@(_AssemblyPaths)" />
```
3. A single list while avoiding strings:
```xml
<ItemGroup>
  <_AssemblyPaths Include="path/to/Asm1.dll">
    <clearinitlocals>true</clearinitlocals>
    <beforefieldinit>false</beforefieldinit>
  </_AssemblyPaths>
</ItemGroup>
<ILLink AssemblyPaths="@(_AssemblyPaths)" />
```
I think this also allows the alternate syntax:
```xml
<ItemGroup>
  <_AssemblyPaths Include="path/to/Asm1.dll" ClearInitLocals="true" />
</ItemGroup>
<ILLink AssemblyPaths="@(_AssemblyPaths)" />
```

My preference would be 3, since SDK components already need to ensure that they pick the right AssemblyPaths, so metadata on that input seems like a logical place for it to go. Using separate boolean metadata options seems consistent with the booleans for global optimizations.

(as an aside, MSBuild is case-insensitive, but I'm going to switch the doc strings to use Pascal casing for the metadata since that seems to be common practice).

I can't think of a good way to combine the global switches with the per-assembly optimizations like we do on the command-line.

Reply:
I implemented option 3 - please let me know if you prefer something else.

Reply:
Options 3 looks fine but I'm not sure how this will be used in practice when the assemblies full paths are resolved in other targets.

I'd also expect we use the same syntax for all other per-assembly settings probably only `-p ACTION ASM` for now, right?

Reply:
That depends:
- If each SDK component owns a different version of the assembly resolution, it could set the optimizations at the same time, depending on whether the assembly was resolved from a runtime pack, nuget package, etc...
- If the assembly resolution from the core SDK is shared by all components, but each SDK component may have different optimizations, then the existing SDK ItemGroups like `ResolvedFileToPublish` can be given metadata for the optimizations. I think that ItemGroup also has information about how the files were resolved that could be used to make the choice.

Yeah, we use it for actions, and I expect we would do the same for other per-assembly settings.